### ISSUE TITLE
1.1.1: Can't connect between s_server and s_client with PSK

### ISSUE THREAD

**Issue by Morten242** [openssl/openssl#6197]
Like the title says: if you try to connect to an instance of s_server from s_client with PSK it doesn't work, failing with ``32676:error:14201076:SSL routines:tls_choose_sigalg:no suitable signature algorithm:..\ssl\t1_lib.c:2484``

It also fails when attempting to use PSK from code, but this is a nice way to provide a reproducer.

Commands run:
```
Console#1>> openssl.exe s_server -nocert -psk 123abc
Console#2>> openssl.exe s_client -psk 123abc
```

**Comment by mattcaswell** [comment]
This is a known issue that we don't currently plan to fix (there is no easy fix).

The problem is that 1.1.1 defaults to using TLSv1.3. PSKs work quite differently in TLSv1.3 compared to TLSv1.2. One area of difference is that PSKs are associated with a hash. The spec requires that if you haven't otherwise specified a hash to use then you should default to SHA-256. The hash for a PSK must match the hash for the negotiated ciphersuite.

The default set of TLSv1.3 ciphersuites that openssl sends, and the selection criteria used means that, in a default configuration, s_server talking to s_client will always select a ciphersuite that uses SHA-384. This means that in the above example the PSK being sent will not be used because it is not compatible with the selected ciphersuite. Since no PSK is available OpenSSL will then attempt to complete a normal handshake. However, since no certificate is being used, this fails.

From a user perspective there are a couple of workarounds for this.
1) You can configure the ciphersuites so that a ciphersuite compatible with the PSK is used, e.g. by adding "-ciphersuites TLS_AES_128_GCM_SHA256" onto the end of your s_client or s_server command line
2) Use the new psk_session option to s_client/s_server which associates the PSK value to a given hash

From an openssl perspective there are a few things we could do to try and fix this:
1) Ignore the "should" in the spec so that the default hash used matches the default ciphersuite used. We don't like this idea. The should is there for a reason, and we shouldn't go off and do our own thing unless we've got a really good reason.
2) We could change the selection criteria so that by default we choose a ciphersuite that uses SHA-256. We don't like this idea either because it has implications beyond PSKs (it would affect uses where PSKs are not in use) and it has security impacts
3) We could make our selection criteria more intelligent so that if (and only if) a PSK has been offered by the client then we should prioritise ciphersuites that match the hash size for that PSK. This adds significant complexity to our ciphersuite selection code. At the point where ciphersuites are selected we haven't chosen a PSK yet - so we would have to leap ahead in our processing and go and peek at what PSKs are there in order to influence the ciphersuite selection choice. This would add significant complexity. It's also not clear to me whether that would actually always be desirable, e.g. if a PSK is available but we're also able to negotiate a normal connection without using that PSK, and that normal connection would use a "stronger" ciphersuite - what should we do?

None of the possible solutions are very appealing - so at the moment the view is that doing nothing is the best option. The possible fixes are all horrible and the user workaround is straight forward.

**Comment by richsalz** [comment]
Nice writeup, @mattcaswell .  Should some edited version of the *first* paragraphs go into the docs?  (Not the list of why the various attempts to fix are bad.)

**Comment by mattcaswell** [comment]
Yes, that's probably a good idea. We should leave this issue open until that is done.

**Comment by Morten242** [comment]
Thanks for the reply! So the way I understand it now is that it's up to users of OpenSSL to specify a matching hash algorithm and/or ciphersuite if they're setting up a connection using PSK?

OT: This seems like a somewhat strange requirement from the RFC.

**Comment by kroeckx** [comment]
The `-psk_session` option is also undocumented in the manpages.

**Comment by mattcaswell** [comment]
I've been thinking this one over and I've come up with an alternative fix that might be good enough (and is at least better than what we've got).

Previously I suggested this:

> We could make our selection criteria more intelligent so that if (and only if) a PSK has been offered by the client then we should prioritise ciphersuites that match the hash size for that PSK. 

The problem with this is that it adds quite a lot of complexity. However I think we could implement a weaker form of this check. It is simple to check whether a server has been configured with a certificate or not. Without one, it is only possible to connect via PSK. Also, if we are using the "old style" (i.e. OpenSSL 1.1.0 and below) callbacks, then we know that the PSK will default to using SHA-256. Therefore if both of those conditions hold then we could prefer SHA-256 during the ciphersuite selection.

We don't actually *know* what the client has offered at that point. E.g. there could be a problem if we have configured both "old style" *and* "new style" callbacks - in that case we don't actually know that the PSK will default to SHA-256. But I think that issue can probably be ignored for now.

PR coming up.
